File: backend/app/services/curve_dictionary.py
import re
from dataclasses import dataclass


@dataclass(frozen=True)
class CurveDefinition:
    key: str
    family: str
    label: str
    unit: str
    color: str
    mnemonics: tuple[str, ...]


CURVE_DEFINITIONS: tuple[CurveDefinition, ...] = (
    CurveDefinition("gamma", "gamma-ray", "Natural Gamma", "API", "#ef4444", ("NG", "NGAM", "GR", "GAMMA", "CGR", "SGR")),
    CurveDefinition("resistivity", "resistivity", "Resistivity", "ohm.m", "#2563eb", ("RS", "RES", "RESD", "RESS", "HRD", "SPR", "16N", "64N")),
    CurveDefinition("density", "density", "Density", "g/cc", "#16a34a", ("DENS", "DEN", "RHOB", "LSD")),
    CurveDefinition("caliper", "caliper", "Caliper", "mm", "#d97706", ("CL", "CAL", "CALI", "CALP", "CALIPER")),
    CurveDefinition("sp", "spontaneous-potential", "Spontaneous Potential", "mV", "#7c3aed", ("SP",)),
    CurveDefinition("point_resistance", "point-resistance", "Point Resistance", "ohm.m", "#f97316", ("PR",)),
    CurveDefinition("bed_resolution_density", "bed-resolution-density", "Bed Resolution Density", "cps", "#22c55e", ("BD",)),
    CurveDefinition("neutron", "neutron", "Neutron", "cps", "#84cc16", ("NN",)),
    CurveDefinition("inclination", "deviation", "Inclination", "deg", "#0f766e", ("DV", "INC", "INCL", "INCLINATION")),
    CurveDefinition("azimuth", "azimuth", "Azimuth", "deg", "#0891b2", ("AZ", "AZIM", "AZI", "AZIMUTH")),
    CurveDefinition("sonic", "sonic", "Sonic", "usec", "#a855f7", ("TT", "DT", "PDEL", "SVEL")),
)

_BY_MNEMONIC = {
    mnemonic: definition
    for definition in CURVE_DEFINITIONS
    for mnemonic in definition.mnemonics
}


def curve_definition_for(mnemonic: str, description: str = "") -> CurveDefinition | None:
    code = mnemonic.upper()
    text = f"{code} {description}".upper()
    if code in {"DEPT", "DEPTH", "MD"}:
        return CurveDefinition("depth", "depth", "Depth", "m", "#94a3b8", ("DEPT", "DEPTH", "MD"))
    if definition := _BY_MNEMONIC.get(code):
        return definition
    if "GAMMA" in text:
        return _BY_MNEMONIC["GAMMA"]
    if "RESIST" in text:
        return _BY_MNEMONIC["RES"]
    if "DENS" in text:
        return _BY_MNEMONIC["DENS"]
    if code.startswith("TT"):
        return _BY_MNEMONIC["TT"]
    return None


def normalize_curve_key(mnemonic: str, description: str = "") -> str:
    definition = curve_definition_for(mnemonic, description)
    if definition is not None:
        return definition.key
    clean = re.sub(r"[^a-z0-9]+", "_", mnemonic.lower()).strip("_")
    return clean or "curve"


def curve_presentation(mnemonic: str, unit: str, description: str = "") -> tuple[str, str, str, dict]:
    definition = curve_definition_for(mnemonic, description)
    if definition is None:
        return description or mnemonic, unit, "#64748b", {
            "curve_family": "unmapped",
            "mapping_status": "unmapped",
        }
    return definition.label, unit or definition.unit, definition.color, {
        "curve_family": definition.family,
        "mapping_status": "mapped",
        "canonical_key": definition.key,
    }

File: frontend/src/workbench/data/curveDictionary.ts
import type { Curve } from "../../api/types";

const FAMILY_ORDER = [
  "gamma-ray",
  "resistivity",
  "density",
  "caliper",
  "spontaneous-potential",
  "deviation",
  "azimuth",
  "sonic",
  "bed-resolution-density",
  "neutron",
  "unmapped",
];

const FAMILY_LABELS: Record<string, string> = {
  "gamma-ray": "Gamma ray",
  resistivity: "Resistivity",
  density: "Density",
  caliper: "Caliper",
  "spontaneous-potential": "SP",
  deviation: "Deviation",
  azimuth: "Azimuth",
  sonic: "Sonic",
  "bed-resolution-density": "Bed density",
  neutron: "Neutron",
  unmapped: "Unmapped",
};

const FAMILY_BY_LEGACY_KEY: Record<string, string> = {
  gamma: "gamma-ray",
  ngam: "gamma-ray",
  ngamma: "gamma-ray",
  gr: "gamma-ray",
  resistivity: "resistivity",
  res: "resistivity",
  rs: "resistivity",
  spr: "resistivity",
  density: "density",
  dens: "density",
  rhob: "density",
  caliper: "caliper",
  calp: "caliper",
  calp_incl: "caliper",
  sp: "spontaneous-potential",
  inclination: "deviation",
  incl: "deviation",
  azimuth: "azimuth",
  az: "azimuth",
};

const MNEMONIC_BY_LEGACY_KEY: Record<string, string> = {
  gamma: "GR",
  ngam: "NGAM",
  ngamma: "NGAM",
  res: "RES",
  spr: "SPR",
  dens: "DENS",
  calp: "CALP",
  calp_incl: "CALP/INCL",
  sp: "SP",
};

export function curveFamily(curve: Curve) {
  const family = curve.curve_metadata?.curve_family;
  if (typeof family === "string" && family.trim()) return family;
  return FAMILY_BY_LEGACY_KEY[curve.key.toLowerCase()] ?? familyFromLabel(curve.label);
}

export function curveFamilyLabel(curve: Curve) {
  const family = curveFamily(curve);
  return FAMILY_LABELS[family] ?? family.replaceAll("-", " ");
}

export function curveMappingStatus(curve: Curve) {
  const status = curve.curve_metadata?.mapping_status;
  if (typeof status === "string" && status.trim()) return status;
  return curveFamily(curve) === "unmapped" ? "unmapped" : "mapped-by-key";
}

export function curveMnemonic(curve: Curve) {
  const mnemonic = curve.curve_metadata?.mnemonic;
  if (typeof mnemonic === "string" && mnemonic.trim()) return mnemonic;
  return MNEMONIC_BY_LEGACY_KEY[curve.key.toLowerCase()] ?? curve.key;
}

export function compareCurvesByFamily(left: Curve, right: Curve) {
  const leftRank = FAMILY_ORDER.indexOf(curveFamily(left));
  const rightRank = FAMILY_ORDER.indexOf(curveFamily(right));
  const normalizedLeftRank = leftRank === -1 ? FAMILY_ORDER.length : leftRank;
  const normalizedRightRank = rightRank === -1 ? FAMILY_ORDER.length : rightRank;
  if (normalizedLeftRank !== normalizedRightRank) return normalizedLeftRank - normalizedRightRank;
  return left.label.localeCompare(right.label);
}

function familyFromLabel(label: string) {
  const text = label.toLowerCase();
  if (text.includes("gamma")) return "gamma-ray";
  if (text.includes("resist")) return "resistivity";
  if (text.includes("dens")) return "density";
  if (text.includes("caliper")) return "caliper";
  if (text.includes("spontaneous")) return "spontaneous-potential";
  if (text.includes("incl")) return "deviation";
  if (text.includes("azimuth")) return "azimuth";
  return "unmapped";
}
